src/settings/project/mod.rs

```diff
@@ -143,144 +139,116 @@ impl Manifest {
+                "The --release flag is not compatible with use of the workers_dot_dev field"
+            )
+        }
 
-        let webpack_config = match environment {
-            Some(environment) => match &environment.webpack_config {
-                Some(webpack_config) => Some(webpack_config.clone()),
-                None => self.webpack_config.clone(),
-            },
-            None => self.webpack_config.clone(),
-        };
+        if release {
+            message::warn("--release will be deprecated");
+        }
 
-        let zone_id = match environment {
-            Some(environment) => match &environment.zone_id {
-                Some(zone_id) => Some(zone_id.clone()),
-                None => self.zone_id.clone(),
-            },
-            None => self.zone_id.clone(),
+        let mut target = Target {
+            project_type: self.project_type.clone(),     // MUST inherit
+            account_id: self.account_id.clone(),         // MAY inherit
+            webpack_config: self.webpack_config.clone(), // MAY inherit
+            zone_id: self.zone_id.clone(),               // MAY inherit
+            workers_dot_dev: true,                       // MAY inherit,
+            // importantly, the top level name will be modified
+            // to include the name of the environment
+            name: self.name.clone(),                   // MAY inherit
+            kv_namespaces: self.kv_namespaces.clone(), // MUST NOT inherit
+            route: None,                               // MUST NOT inherit
+            routes: self.routes.clone(),               // MUST NOT inherit
         };
 
-        let project_type = self.project_type.clone();
+        let environment = self.get_environment(environment_name)?;
+        let (route, workers_dot_dev) = self.negotiate_with_terrorists(environment, release)?;
+        target.route = route;
+        target.workers_dot_dev = workers_dot_dev;
+        if let Some(environment) = environment {
+            target.name = if let Some(name) = &environment.name {
+                name.clone()
+            } else {
+                match environment_name {
+                    Some(environment_name) => format!("{}-{}", self.name, environment_name),
+                    None => failure::bail!("You must specify `name` in your wrangler.toml"),
+                }
+            };
+            if let Some(account_id) = &environment.account_id {
+                target.account_id = account_id.clone();
+            }
+            if environment.routes.is_some() {
+                target.routes = environment.routes.clone();
+            }
+            if environment.webpack_config.is_some() {
+                target.webpack_config = environment.webpack_config.clone();
+            }
+            if environment.zone_id.is_some() {
+                target.zone_id = environment.zone_id.clone();
+            }
+            target.kv_namespaces = environment.kv_namespaces.clone();
```

Comment:
Do we want to wrap `environment.kv_namespaces` in the same conditionals as above? I would think kv inheritance would work the same as other toml fields...

Reply:
@ashleymichal and I talked about this and said that it doesn't really make sense to inherit kv namespaces because it's an anti-pattern to use multiple namespaces across different environments. I don't want my staging worker to use my production kv namespaces. In this case - if you wanted to use the same namespace for both, you would just have to copy + paste to explicitly use the same namespaces for both environments. Open to discussion about this.

Reply:
This makes sense. It could be worth adding this as a comment to clarify why this logic diverges from the rest of the logic preceding it. Other than that, LGTM!